ic-agent/src/agent/agent_test.rs

```diff
@@ -236,3 +237,138 @@ fn status_error() -> Result<(), AgentError> {
+    0, 0, 0, 0, 32, 0, 0, 1, 1, 74, 0, 0, 0, 0, 0, 47, 255, 255, 1, 1, 131, 2, 74, 112, 117, 98,
+    108, 105, 99, 95, 107, 101, 121, 130, 3, 88, 133, 48, 129, 130, 48, 29, 6, 13, 43, 6, 1, 4, 1,
+    130, 220, 124, 5, 3, 1, 2, 1, 6, 12, 43, 6, 1, 4, 1, 130, 220, 124, 5, 3, 2, 1, 3, 97, 0, 153,
+    51, 225, 248, 158, 138, 60, 77, 127, 220, 204, 219, 213, 24, 8, 158, 43, 212, 216, 24, 10, 38,
+    31, 24, 217, 194, 71, 165, 39, 104, 235, 206, 152, 220, 115, 40, 163, 152, 20, 168, 249, 17, 8,
+    106, 29, 213, 12, 190, 1, 94, 42, 83, 183, 191, 120, 181, 82, 136, 137, 61, 170, 21, 195, 70,
+    100, 14, 136, 49, 215, 42, 18, 189, 237, 217, 121, 210, 132, 112, 195, 72, 35, 184, 209, 195,
+    244, 121, 93, 156, 57, 132, 162, 71, 19, 46, 148, 254, 130, 4, 88, 32, 153, 111, 23, 187, 146,
+    107, 227, 49, 87, 69, 222, 167, 40, 32, 5, 167, 147, 181, 142, 118, 175, 235, 93, 67, 209, 162,
+    140, 226, 157, 45, 21, 133, 131, 2, 68, 116, 105, 109, 101, 130, 3, 73, 149, 184, 170, 192,
+    228, 237, 162, 234, 22, 105, 115, 105, 103, 110, 97, 116, 117, 114, 101, 88, 48, 172, 233, 252,
+    221, 155, 201, 119, 224, 93, 99, 40, 248, 137, 220, 78, 124, 153, 17, 76, 115, 122, 73, 70, 83,
+    203, 39, 161, 245, 92, 6, 244, 85, 94, 15, 22, 9, 128, 175, 94, 173, 9, 138, 204, 25, 80, 16,
+    178, 247,
+];
+
+#[test]
+// asserts that a delegated certificate with correct /subnet/<subnetid>/canister_ranges
+// passes the certificate verification
+fn check_subnet_range_with_valid_range() -> Result<(), AgentError> {
+    let _read_mock = mock(
+        "POST",
+        "/api/v2/canister/ivg37-qiaaa-aaaab-aaaga-cai/read_state",
+    )
+    .with_status(200)
+    .with_body(&REQ_WITH_DELEGATED_CERT_RESPONSE)
+    .create();
+    let agent = Agent::builder()
+        .with_transport(ReqwestHttpReplicaV2Transport::create(
+            &mockito::server_url(),
+        )?)
+        .build()?;
+    let runtime = tokio::runtime::Runtime::new().expect("Unable to create a runtime");
+    let result = runtime.block_on(async {
+        agent
+            .read_state_raw(
+                vec![REQ_WITH_DELEGATED_CERT_PATH
+                    .iter()
+                    .map(Label::from)
+                    .collect()],
+                Principal::from_text(REQ_WITH_DELEGATED_CERT_CANISTER).unwrap(),
+            )
+            .await
+    });
+    match result {
+        Ok(_) => Ok(()),
+        Err(other) => panic!("found something unexpected: {:?}", other),
+    }
+}
+
+#[test]
+// asserts that a delegated certificate with /subnet/<subnetid>/canister_ranges that don't include
+// the canister gets rejected by the cert verification because the subnet is not authorized to
+// respond to requests for this canister. We do this by using a correct response but serving it
+// for the wrong canister, which a malicious node might do.
+fn check_subnet_range_with_unauthorized_range() -> Result<(), AgentError> {
```

Comment:
How about a test that checks that verification fails if `/subnet/<subnetid>/canister_ranges` is missing? (I think the code _does_ fail, but still)

Reply:
was a huge pain, but I managed to create a response to add this test